File: python/linked_list/problems/remove_nth_node_from_end.py
from typing import Optional

from linked_list.list_node import ListNode


def remove_nth_node_from_end(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    if not head:
        return None

    dummy: Optional[ListNode] = ListNode(0, head)
    right: Optional[ListNode] = head

    while n > 0:
        right = right.next if right else None
        n -= 1

    left: Optional[ListNode] = dummy
    while right:
        left = left.next if left else None
        right = right.next

    if left and left.next:
        left.next = left.next.next

    return dummy.next

File: src/main/java/io/github/willslaunwhite/linkedlist/problems/RemoveNthNodeFromEnd.java
package io.github.willslaunwhite.linkedlist.problems;

import io.github.willslaunwhite.linkedlist.ListNode;

public class RemoveNthNodeFromEnd {
    public ListNode removeNthNodeFromEnd(ListNode head, int n) {
        if (head == null || head.next == null) {
            return null;
        }

        ListNode right = head;
        ListNode dummy = new ListNode(0, head);
        while (n > 0) {
            right = right.next;
            n--;
        }

        ListNode left = dummy;
        while (right != null) {
            right = right.next;
            left = left.next;
        }
        left.next = left.next.next;

        return dummy.next;
    }
}